Add play-with-number to codelab contents

quwen opened the pull request.

quwen committed:
First commit

quwen committed:
Add test for prime number determiner

mi-bear commented on play-with-number/main.go:
```diff
@@ -0,0 +1,14 @@
+	var num int
+	fmt.Scan(&num)
+
+	fmt.Printf("この数字は%s\n", primenumdeterminer(num))
```
好み的には、
PrimeNumdeterminer
が見やすくて好きかもです。(Global にしちゃう)

mi-bear commented on play-with-number/judge_test.go:
```diff
@@ -0,0 +1,37 @@
+	tests := []struct {
+		name       string
+		args       args
+		wantResult string
```
`want` のままでもいいかな〜と思ったりしました。

mi-bear commented on play-with-number/judge_test.go:
```diff
@@ -0,0 +1,37 @@
+import "testing"
+
+func Test_primenumdeterminer(t *testing.T) {
+	type args struct {
```
`args` を `struct` にするのは理由があるのでしたっけ。

quwen replied:
これ、VSCodeのスニペットなのです。そういうもんなのかなと思って直さなかったのですがこのコードだと `struct` にする必要ないですね。

mi-bear replied:
そうですね、必要のないもの = 説明がつかないもの、になっちゃうから、なくても平気！だと思います〜

quwen committed:
Change variable and function names

mi-bear commented:
Slack に思い出して書いちゃったのも含めてコメントしておきます。

- [x] flag を利用したい。(引数で、game=`素数` を選べるようにして、ゲームスタート！だと楽しい + 拡張性ありそう)
- [x] time パッケージを使って、数字に変換したい。(今日は素数日？とか)
- [x] go.mod 入れたい。

quwen committed:
Add go.mod

quwen committed:
Add option

mi-bear commented on play-with-number/main.go:
```diff
@@ -0,0 +1,31 @@
+	}
+}
+
+func init() {
```
init は main  よりも上に合ったほうがわかりやすいかも。

mi-bear replied:
マストではないけど、example みたいに、
const で、デフォルト値、ヘルプの文章を定義しておくと良さそう。(init の外でいい)

```
const (
  default = ""
  usage = "ゲームのメニューを選択" 
)
var game string
```

mi-bear commented on play-with-number/main.go:
```diff
@@ -0,0 +1,31 @@
+	// TODO:timeを使って今日の日付を数字に変換する
+
+	if game != "" {
+		fmt.Print("数字を入力してください\n")
```
標準入力で数字の入力を求める処理を prime という関数に分けちゃうのはどうでしょう。


switch で、 `game == "prime"` ならば、prime という関数を呼び出す。
`game == ""` だったら、エラーメッセージ。

というふうにすると、拡張性がありそう。

quwen committed:
Add init() constants

quwen committed:
Change args from struct to int.

quwen committed:
Add prime()

quwen commented:
> * [ ]  time パッケージを使って、数字に変換したい。(今日は素数日？とか)

日付の数値化って、例えば4月５日だったら `45` か`405`のどちらになるのでしょうか？

mom0tomo commented:
日付のイメージは、私はこんな感じでした
2021/4/5 → 20,210,405

quwen committed:
Add judgement using todays date.

quwen committed:
Add menu 'today'.

quwen commented:
下記に変更
main.go

1. 引数を取得する
2. switch でゲームふりわけ
3. primeNumberDeterminer() <- 対話して、Prime() を呼んだ結果を表示する
4. todaysDateDeterminer() <- 日付を数字に変換して、Prime() を呼んだ結果を表示する

一度書いてみて、3, 4, がある程度共通化できるかもしれない。

---
以下を追加

- [x] prime_test.go = Prime() のテスト
- [x] Prime()

quwen committed:
Add package prime.

quwen committed:
Remove judge.go.

mi-bear commented on play-with-number/main.go:
```diff
@@ -0,0 +1,72 @@
+	flag.StringVar(&game, "game", defaultGame, usage)
+}
+
+const (
```
const は init で利用しているので上にあったほうが見やすいな〜！って思いました。

mi-bear commented on play-with-number/main.go:
```diff
@@ -0,0 +1,72 @@
+	"github.com/WomenWhoGoTokyo/codelab/play-with-number/prime"
+)
+
+var game string
```
```
var (
	game string
	num int
)
```
このように書けますよ！

mi-bear replied:
ただし、下で説明しているとおり、ここで定義している `num` は必要なさそうです。

mi-bear commented on play-with-number/main.go:
```diff
@@ -0,0 +1,72 @@
+	switch game {
+
+	case "prime":
+		getNum()
```
ここで、グローバル変数的に上部で指定している、var num int に値を入れていると思うのですが、ここへの書き込みはこの箇所だけで、ちょっと見通しが良くないかもしれません。


しかも、`getNum()` 関数は戻り値も設定されている...!!! 

なので、

```
fmt.Printf("この数字は%s\n", PrimeNumdeterminer(
			getNum(),
		))
```

これで良さそうです。(特に改行する必要はないです)

mi-bear replied:
もしくは、getNum は特につくらず、case "prime" の処理の中で、標準入力の処理を入れてもシンプルだと思いました。

mi-bear commented on play-with-number/main.go:
```diff
@@ -0,0 +1,72 @@
+func getNum() int {
+	fmt.Print("数字を入力してください\n")
+
+	fmt.Scan(&num)
```
ここでは、こうなるかな...?
```
var num int
fmt.Scan(&num)
return num
```

mi-bear commented on play-with-number/main.go:
```diff
@@ -0,0 +1,72 @@
+func TodaysDatedeterminer() (result string) {
+
+	t := time.Now()
+	const layout = "20060102"
```
ここを const にしているのはなぜなのでしたっけ。

quwen replied:
例が全部constだったからそうなのか、と思ったのですが、go docにそんな事書いてなかったし、varでやっても動いたのでvarにします。

mi-bear commented on play-with-number/main.go:
```diff
@@ -0,0 +1,72 @@
+	const layout = "20060102"
+
+	today, err := strconv.Atoi(t.Format(layout))
+	if err != nil {
```
ここを Fatal で終わらせるのではなく、error を返却して main で処理してあげたいです。

mi-bear commented on play-with-number/main.go:
```diff
@@ -0,0 +1,72 @@
+	return num
+}
+
+func PrimeNumdeterminer(num int) (msg string) {
```
mst という変数は利用していないので、戻り値は `string` のままで大丈夫です〜

quwen committed:
Rename todaysDateDeterminer() -> convertTodayNum().
Change functions in main to private.
Improved variable passing to be simpler.
Add error handling and comment.
Make the output of prime number determination result common.

mi-bear commented on play-with-number/prime/prime.go:
```diff
@@ -0,0 +1,18 @@
+package prime
+
+func Prime(n int) bool {
+	var divisor []int
```
これ、数の上限がわかっているので、容量を指定して作っちゃってもいいかもです。

mi-bear commented on play-with-number/prime/prime_test.go:
```diff
@@ -0,0 +1,30 @@
+
+	tests := []struct {
+		name string
+		args int
```
引数は複数形ではないので `arg` で良さそう！と思いました。

mi-bear commented on play-with-number/prime/prime_test.go:
```diff
@@ -0,0 +1,30 @@
+func TestPrime(t *testing.T) {
+
+	tests := []struct {
+		name string
```
name で返却するものはないので、必要ないかも？

mi-bear reviewed (commented):
prime の部分にコメントを入れました！
あと、main.go の日本語コメントは、miki さんに伝える用だったので、書かなくて良いやつです〜

quwen committed:
Remove comments
Change variable args -> arg.
Chabge to specify the capacity of the divisor.

mi-bear commented:
LGTM ฅʕ•ᴥ•ʔฅ マージするね！
